auth/src/lib.rs

```diff
@@ -0,0 +1,193 @@
+#![no_std]
+
+multiversx_sc::imports!();
+
+mod constants;
+
+use crate::constants::{ProofData, TransferData, OLD_KEY_RETENTION, Operator};
+use core::ops::Deref;
+use multiversx_sc::api::{ED25519_SIGNATURE_BYTE_LEN, KECCAK256_RESULT_LEN};
+
+#[multiversx_sc::contract]
+pub trait Auth {
+    #[init]
+    fn init(&self, recent_operators: MultiValueEncoded<TransferData<Self::Api>>) {
+        for operator in recent_operators.into_iter() {
+            self.transfer_operatorship(operator);
+        }
+    }
+
+    #[endpoint(validateProof)]
+    fn validate_proof(
+        &self,
+        message_hash: ManagedByteArray<KECCAK256_RESULT_LEN>,
+        proof_data: ProofData<Self::Api>,
+    ) -> bool {
+        let operators_hash = self.get_operators_hash(
+            &proof_data.operators,
+            &proof_data.weights,
+            &proof_data.threshold,
+        );
+
+        let operators_epoch_mapper = self.epoch_for_hash(&operators_hash);
+        let epoch = self.current_epoch().get();
+
+        require!(
+            !operators_epoch_mapper.is_empty()
+                && epoch - operators_epoch_mapper.get() < OLD_KEY_RETENTION,
```

Comment:
You are using `operators_epoch_mapper.get()` two times. Better assign a variable. This way the SC won't make two storage accesses and save gas.